wkcuber/tile_cubing.py

```diff
@@ -1,60 +1,233 @@
+                            )
+                            x_max = (
+                                x_max if x_max > coordinate_value else coordinate_value
+                            )
+                        elif occurrence[1] == "y":
+                            occurrence_end_index = (
+                                occurrence_begin_index + y_decimal_length - y
+                            )
+                            index_offset_caused_by_brackets_and_specific_length += y
+                            coordinate_value = int(
+                                file[occurrence_begin_index:occurrence_end_index]
+                            )
+                            y_min = (
+                                y_min
+                                if y_min and y_min < coordinate_value
+                                else coordinate_value
+                            )
+                            y_max = (
+                                y_max if y_max > coordinate_value else coordinate_value
+                            )
+                        else:
+                            occurrence_end_index = (
+                                occurrence_begin_index + z_decimal_length - z
+                            )
+                            index_offset_caused_by_brackets_and_specific_length += z
+                            coordinate_value = int(
+                                file[occurrence_begin_index:occurrence_end_index]
+                            )
+                            z_min = (
+                                z_min
+                                if z_min and z_min < coordinate_value
+                                else coordinate_value
+                            )
+                            z_max = (
+                                z_max if z_max > coordinate_value else coordinate_value
+                            )
+
+    return z_min, z_max, y_min, y_max, x_min, x_max, arbitrary_file, file_count
 
-def tile_cubing_job(target_wkw_info, z_batches, source_path, batch_size, tile_size):
+
+def find_file_with_dimensions(
+    file_path_pattern: str,
+    x_value: int,
+    y_value: int,
+    z_value: int,
+    x_decimal_length: int,
+    y_decimal_length: int,
+    z_decimal_length: int,
+) -> str:
+    # optimize the bounds
+    x_missing_number_length_offset = floor(log10(max(x_value, 1))) + 1
+    y_missing_number_length_offset = floor(log10(max(y_value, 1))) + 1
+    z_missing_number_length_offset = floor(log10(max(z_value, 1))) + 1
+    upper_bound_x = x_decimal_length - x_missing_number_length_offset
+    upper_bound_y = y_decimal_length - y_missing_number_length_offset
+    upper_bound_z = z_decimal_length - z_missing_number_length_offset
+
+    # try to find the file with all combinations of number lengths
+    for z_missing_number_length in range(upper_bound_z):
+        for y_missing_number_length in range(upper_bound_y):
+            for x_missing_number_length in range(upper_bound_x):
```

Comment:
So, with the default pattern of `{zzzzzzzzzz}/{yyyyyyyyyy}/{xxxxxxxxxx}`, this method will do 10^3 = 1000 iterations to check whether a certain path exists? And this will be done for files? This seems quite inefficient to me. Why is it necessary to support every combination of padding? Can't we simply assume that the user either gives a pattern like `{z}/{y}/{x}` (single digit --> no padding for all indices) or a pattern like `{zzz}/{yyy}/{xxx}` (multiple digit --> precise padding to create a length of three). Then, we would only do one file access per tile instead of 1000?

I think it's a fair assumption that the user-provided data has to be consistent (i.e., the padding is consistent or there is no padding).

Reply:
Yes that's true. I could not come up with a regex that works with glob and lets you look for numbers with a dynamic length.

But since Norman wanted everything to be backwards compatible this is the solution I came up with. Please take a look at the tests. The data set does not meet the criteria, that all digits have the same length. 
Do you have a better solution that is backwards compatible?

Reply:
> But since Norman wanted everything to be backwards compatible this is the solution I came up with. Please take a look at the tests. The data set does not meet the criteria, that all digits have the same length.

Which test case are you referring to? From what I see, all the test data, uses an unpadded naming schema. This means, if an index is `5`, the file name is `5.tiff` and if the index is `511`, we know it's `511.tiff`. The provided pattern would simply be `{x}/...`. If the data is padded like `005.tiff` and `511.tiff`, the pattern should be `{xxx}`.
I think, only supporting these two options (consistent padding or consistent non-padding) is fair. I'll ask Norman about :)
Still, to not break the interface, we would need to recognize which of the two modes the user provided (padded or unpadded). I'll think about this.

> I could not come up with a regex that works with glob and lets you look for numbers with a dynamic length.

Maybe we shouldn't use glob then? We could list all files and apply the regular expression by ourselves. But let's discuss this in person, maybe.

Reply:
I talked to Norman and we agreed on going with "either unpadded or consistently padded". So, you don't need to iterate through all combinations and can simply do two file operations per look up. Alternatively, one could traverse over the provided path, list all files and match them against a regexp, but one would need to be careful to not traverse unrelated folders. To get this issue done, the first solution is more sensible :)

Can you implement this please? Will re-review again afterwards :)